README.md
**Notice:** This repo was a clone of the one on [Dominic Radermacher's website](https://mockmoon-cybernetics.ch/computer/p-touch2430pc/), and he is the original author. The original is available at (https://github.com/clarkewd/ptouch-print/), with his permission, to help increase visibility of the project and to encourage collaboration and further development.


# ptouch-print

ptouch-print is a command line tool to print labels on Brother P-Touch
printers on macOS.

There is no need to install the printer via CUPS, the printer is accessed
directly via libusb.

The tool was written for and tested with the `PT-2730`, but it should also
work with the `PT-2430PC` and `PT-1230PC` (untested so far). Maybe others work too (please report USB VID and PID so I can include support for further models, too).

**Models in `libptouch.c` include:**

- PT-2420PC
- PT-1230PC
- PT-2430PC
- PT-2730
- PT-P700 ( see note below about mode switch )
- PT-D450

**Further info can be found at:**
https://mockmoon-cybernetics.ch/computer/p-touch2430pc/

### Compile instructions:

  The project is self-contained. There are no additional dependencies.

  Open the Xcode project. Use the menu command **Product** > **Run**

### PT-2730 on Mac:

Brother dropped support for the PT-2730 with the release of macOS Big Sur (version 11.0) in 2020.

I am furious that Brother dropped support. How hard can it be, it's just a simple USB device?

So I brought this Linux command line tool to macOS. (I checked that it wasn't already in macPorts.)

The commandline tool depended on two other libraries: 

* libUSB - so I grabbed the relevant source files, and include them here.

* libgd - a graphics manipulation library. I just re-implemented what I needed from it in terms of 
the Core Graphics framework that ships with macOS.

For fonts, You can use any of the font names in **/Applications/Font Book**. For fonts with variants
in Font Book like **Bold** or **Italic** just append those modifiers. **Helvetica** is the default font.

## example command lines:

`~$ ptouch-print --font "Snell Roundhand Black" --text "Hello World  " --writepng "ptouch.png"`


<span style="color:gray">PT-2730 found on USB bus 4, device 17</span>

<span style="color:gray">choosing font size=95</span>


`~$ ptouch-print --font "Snell Roundhand" --text "Hello World  " --writepng "ptouch1.png"`

<span style="color:gray">PT-2730 found on USB bus 4, device 17</span>

<span style="color:gray">choosing font size=95</span>


And lastly: Here is an actual photo of the output as printed on my label maker:

![](images/print.jpg)

### PT-2730

My **PT-2730** always wants to put 1" of leader on each print. If the print is just simple
text, the **PT-2730** doesn't print the first half of the first character. But if I include a small
.png first, then the whole png and the following text print just fine. I generally test by having
`ptouch-print` write to a png first, as a preview of what will print.

Here is the actual command line I used to print labels for the jars in my refrigerator:

`ptouch-print --image /Users/david/spiral.png --cutmark --text \ Sauerkraut\  --cutmark --text \ Yeast\  --cutmark --text \ Yeast\  --cutmark --text \ Yeast\  `

where `spiral.png` is: ![](images/spiral.png)

multiple lines of text would be:

`ptouch-print --text "line 1" "line 2" "line 3"`

### PT-P700:

- As the PT-P700 does not have a toggle switch for the printer mode, you may need to switch the mode first
- If you plug in the PT-P700 and the green light above the PLite button is on, the printer won't be recognized by `ptouch`. You must hold down the PLite button for ~ 2 seconds and the light will turn off and the system will find the device again.
- This step allowed it to be detected as `vid 0x04F9 pid 0x2061` insetad of `vid 0x04F9 pid 0x2064`
- This allows it to be found, (eg: `./ptouch-print --info`) and also puts it into a mode that it can be interfaced with


### A Note from Dominic's website

"Dear visitor, currently I have absolutely no time for improvements on this project (my free time currently is about one or two hours PER MONTH)..."

### Collaboration

You may still, of course, contact Dominic directly if you'd like. But I think GitHub is easier for collaboration, and he has given me permission to host a clone of his repo here, passing code contributions back to him.

### License

The original ptouch-print is GNU GPL 3, so my work is too. libusb is under the MIT license.



ChangeLog

date		ver	description of changes

2023-12-19	0.0.2	- If an image is too tall for the tape, scale it down
			  uniformly to fit

2023-12-17	0.0.1	- David Phillip Oster’s port to macOS.

2015-12-29	1.3.2	- added inital ptouch-gtk - it is WITHOUT function
			  for now. I just have somewhere to start...

2015-11-12	1.3.1	- small bugfix
			- now using autotools

2015-02-13	1.3	- merged ptouch-printpng and ptouch-printtext to one
			  single tool so that multiple texts and images can be
			  printed in one single pass.
			- added ptouch_cutmark() function
			- added initial gettext() support

2015-01-05	1.2.3	- fixed a bug that did cut off chars which go below the
			  font baseline (like g,q,j,...)
			  Seems that gdImageStringFT() interprets the y coord
			  as baseline of the font (is this the case for all
			  fonts??) instead of the bottom most pixel.

2014-12-20	1.2.2	- added support for printing up to 4 lines of text

2014-11-01	1.2.1	- added support for PT-1230PC printer (not tested)

2014-10-1?	1.2	- added tool 'ptouch-printtext' which can print labels
			  with one or two lines of text

2014-10-11	1.1	- easier adding of (yet) unsupported p-touch printers
			  and tape widths
			- code cleanup

2014-10-08	1.0	Initial release


ptouch-print/gdMac.m
//  gdMac.m
//  ptouch-print
//
//  Created by David Phillip Oster on 12/16/23.
//

#import <AppKit/AppKit.h>

#import <CoreGraphics/CoreGraphics.h>

#import "gd.h"

static NSData *sDashStyle = nil;

int gdFTUseFontConfig(int flag) {
 return GD_TRUE;
}

int gdImageBlue(gdImage *im, int index) {
 return gdImageGetPixel(im, index, 0);
}

// Since we are working in grayscale for black and white, just use the average.
int gdImageColorAllocate(gdImage *im, int red, int green, int blue) {
  return (red + green + blue)/3;
}

void gdImageCopy(gdImage * dst, gdImage * src, int dstX, int dstY, int srcX, int srcY, int w, int h) {
  CGRect srcR = CGRectMake(srcX, srcY, w, h);
  CGImageRef srcImage = CGBitmapContextCreateImage(src);
  CGImageRef srcRectImage = CGImageCreateWithImageInRect(srcImage, srcR);
  CGRect destR = CGRectMake(dstX, dstY, w, h);
  CGContextDrawImage(dst, destR, srcRectImage);
  CGImageRelease(srcRectImage);
  CGImageRelease(srcImage);
}

gdImage *gdImageCreateFromPng(FILE *inFile) {
  fseek(inFile, 0, SEEK_END);
  NSInteger length = ftell(inFile);
  rewind(inFile);
  NSMutableData *data = [[NSMutableData alloc] initWithLength:length];
  if (data && 1 == fread(data.mutableBytes, data.length, 1, inFile)) {
    NSImage *image = [[NSImage alloc] initWithData:data];
    if (image) {
      CGRect r = CGRectMake(0, 0, image.size.width, image.size.height);
      CGImageRef ref = [image CGImageForProposedRect:&r context:nil hints:nil];
      if (ref) {
        gdImage *result = gdImageCreatePalette(image.size.width, image.size.height);
        if (result) {
          CGContextDrawImage(result, r, ref);
          return result;
        }
      }
    }
  }
  return NULL;
}

gdImage *gdImageCreatePalette(int x, int y) {
  const int bitsPerPixel = 8;
  static CGColorSpaceRef gray = NULL;
  if(NULL == gray){
    gray = CGColorSpaceCreateWithName(kCGColorSpaceGenericGray);
  }
  uint32_t info = kCGImageByteOrderDefault | kCGImageAlphaNone;
  CGContextRef result = CGBitmapContextCreate(NULL, x, y, bitsPerPixel, (x*bitsPerPixel + 7)/8, gray, info);
  if (result) {
    uint8_t *p = (uint8_t *)CGBitmapContextGetData(result);
    int rowBytes = (int)CGBitmapContextGetBytesPerRow(result);
    for(int j = 0;j < y; ++j) {
      memset(p, 255, rowBytes);
      p += rowBytes;
    }
  }
  return result;
}

void gdImageDestroy(gdImage *im) {
  CGContextRelease(im);
}

int gdImageGetPixel(gdImage *im, int x, int y) {
  uint8_t *base = (uint8_t *)CGBitmapContextGetData(im);
  return base[x + y*CGBitmapContextGetBytesPerRow(im)];
}

int gdImageGreen(gdImage *im, int index) {
 return gdImageGetPixel(im, index, 0);
}

void gdImageLine(gdImage *im, int x1, int y1, int x2, int y2, int color) {
  CGFloat colors[3];
  CGContextBeginPath(im);
  if (color == gdStyled) {
    if (sDashStyle) {
      colors[0] = 0;    // black
      colors[1] = 1.0;  // opaque
      // phase=1, so we start with length of transparent.
      CGContextSetLineDash(im, 1, sDashStyle.bytes, sDashStyle.length/sizeof(CGFloat));
    }
  } else {
    CGContextSetLineDash(im, 0, NULL, 0);
    colors[0] = color / 255.0;
    colors[1] = 1.0;
  }
  CGContextSetStrokeColor(im, colors);
  // since pen is one pixel wide, offset in X to fill pixel with color
  CGFloat offset = (x1 == x2) ? -0.5 : 0;
  CGContextMoveToPoint(im, x1 + offset, y1);
  CGContextAddLineToPoint(im, x2 + offset, y2);
  CGContextStrokePath(im);
}

int gdImagePng(gdImage *im, FILE *outFile) {
  CGSize cgImageSize;
  cgImageSize.width = CGBitmapContextGetWidth(im);
  cgImageSize.height = CGBitmapContextGetHeight(im);
  CGImageRef cgImage = CGBitmapContextCreateImage(im);
  if (cgImage) {
    NSImage *image = [[NSImage alloc] initWithCGImage:cgImage size:cgImageSize];
    if (image) {
      NSData *imageData = [image TIFFRepresentation];
      if (imageData) {
        NSBitmapImageRep *imageRep = [NSBitmapImageRep imageRepWithData:imageData];
        NSDictionary *imageProps = @{NSImageCompressionFactor: @(1.0)};
        imageData = [imageRep representationUsingType:NSBitmapImageFileTypePNG properties:imageProps];
        if (imageData) {
          return 1 != fwrite(imageData.bytes, imageData.length, 1, outFile);
        }
      }
    }
  }
  return 0;
}

int gdImageRed(gdImage *im, int index) {
 return gdImageGetPixel(im, index, 0);
}

void gdImageSetStyle(gdImage *im, int *style, int styleLength){
  if (styleLength == 6) {
    int i = 0;
    Boolean isDash = YES;
    for (; i < 3 && isDash;++i) {
      isDash |= style[i] == gdTransparent;
    }
    for (; i < 6 && isDash;++i) {
      isDash |= style[i] == 0;
    }
    if (isDash) {
      // input is specified in pixels, one entry per pixel. transparent first. CGContext wants line segments lengths, drawing first.
      CGFloat dash[] = { 3, 3 };
      sDashStyle = [NSData dataWithBytes:dash length:sizeof(CGFloat)];
    }
  }
}

char *gdImageStringFT(gdImage *im, int *brect, int fg, char *fontname, double ptsize, double angle, int x, int y, char *string) {
  @autoreleasepool {
    NSFont *font = [NSFont fontWithName:[NSString stringWithUTF8String:fontname] size:ptsize];
    if (nil == font) {
      return "Can’t get font";
    }
    NSString *nsString = [NSString stringWithUTF8String:string];
    if (nil == nsString) {
      return "Can’t convert to UTF8";
    }
    NSDictionary *dict = @{NSFontAttributeName: font, NSForegroundColorAttributeName : [NSColor colorWithWhite:fg/255.0 alpha:1]};
    if (im) {
      NSGraphicsContext *context = [NSGraphicsContext graphicsContextWithCGContext:im flipped:NO];
      NSGraphicsContext *preserve = [NSGraphicsContext currentContext];
      [NSGraphicsContext setCurrentContext:context];
      [nsString drawAtPoint:CGPointMake(x, gdImageSY(im) - y) withAttributes:dict];
      [NSGraphicsContext setCurrentContext:preserve];
    }
    if (brect) {
      CGRect r = [nsString boundingRectWithSize:CGSizeMake(100000, 100000) options:0 attributes:dict];
      brect[0] = r.origin.x + x; // lower left
      brect[1] = r.origin.y + y + r.size.height;
      brect[2] = r.origin.x + x + r.size.width; // lower right
      brect[3] = r.origin.y + y + r.size.height;

      brect[4] = r.origin.x + x;  // upper left
      brect[5] = r.origin.y + y;
      brect[6] = r.origin.x + x + r.size.width; // upper right
      brect[7] = r.origin.y + y;
    }
    return NULL;
  }
}

int gdImageSX(gdImage *im) {
 return (int) CGBitmapContextGetWidth(im);
}

int gdImageSY(gdImage *im) {
 return (int) CGBitmapContextGetHeight(im);
}



ptouch-print/include/ptouch.h
/*
	ptouch-print - Print labels with images or text on a Brother P-Touch
	
	Copyright (C) 2015-2019 Dominic Radermacher <blip@mockmoon-cybernetics.ch>

	This program is free software; you can redistribute it and/or modify it
	under the terms of the GNU General Public License version 3 as
	published by the Free Software Foundation
	
	This program is distributed in the hope that it will be useful, but
	WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
	See the GNU General Public License for more details.
	
	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software Foundation,
	Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*/

#include <stdint.h>

#ifdef __APPLE__
#include "TargetConditionals.h"
#endif
#if TARGET_OS_MAC // Oster: i.e., is Mac.
#include "libusb.h"
#else
#include <libusb-1.0/libusb.h>
#endif

struct _pt_tape_info {
	uint8_t mm;		/* Tape width in mm */
	uint8_t px;		/* Printing area in px */
};

#define FLAG_NONE		(0x00)
#define FLAG_UNSUP_RASTER	(0x01)
#define FLAG_RASTER_PACKBITS	(0x02)
#define FLAG_PLITE		(0x04)
#define FLAG_P700_INIT		(0x08)

struct _pt_dev_info {
	int vid;		/* USB vendor ID */
	int pid;		/* USB product ID */
	char *name;
	int max_px;		/* Maximum pixel width that can be printed */
	int flags;
};
typedef struct _pt_dev_info *pt_dev_info;

struct __attribute__((packed, aligned(4))) _ptouch_stat {
	uint8_t printheadmark;	// 0x80
	uint8_t size;		// 0x20
	uint8_t brother_code;	// "B"
	uint8_t series_code;	// "0"
	uint8_t model;
	uint8_t country;	// "0"
	uint16_t reserved_1;
	uint16_t error;		// table 1 and 2
	uint8_t media_width;	// tape width in mm
	uint8_t media_type;	// table 4
	uint8_t ncol;		// 0
	uint8_t fonts;		// 0
	uint8_t jp_fonts;	// 0
	uint8_t mode;
	uint8_t density;	// 0
	uint8_t media_len;	// table length, always 0
	uint8_t status_type;	// table 5
	uint8_t phase_type;
	uint16_t phase_number;	// table 6
	uint8_t notif_number;
	uint8_t exp;		// 0
	uint8_t tape_color;	// table 8
	uint8_t text_color;	// table 9
	uint32_t hw_setting;
	uint16_t reserved_2;
};
typedef struct _ptouch_stat *pt_dev_stat;

struct _ptouch_dev {
	libusb_device_handle *h;
	pt_dev_info devinfo;
	pt_dev_stat status;
	uint8_t tape_width_px;
};
typedef struct _ptouch_dev *ptouch_dev;

int ptouch_open(ptouch_dev *ptdev);
int ptouch_close(ptouch_dev ptdev);
int ptouch_send(ptouch_dev ptdev, uint8_t *data, size_t len);
int ptouch_init(ptouch_dev ptdev);
int ptouch_lf(ptouch_dev ptdev);
int ptouch_ff(ptouch_dev ptdev);
int ptouch_eject(ptouch_dev ptdev);
int ptouch_getstatus(ptouch_dev ptdev);
int ptouch_getmaxwidth(ptouch_dev ptdev);
int ptouch_enable_packbits(ptouch_dev ptdev);
int ptouch_rasterstart(ptouch_dev ptdev);
int ptouch_sendraster(ptouch_dev ptdev, uint8_t *data, size_t len);


ptouch-print/ptouch-print.c
/*
	ptouch-print - Print labels with images or text on a Brother P-Touch
	
	Copyright (C) 2015-2019 Dominic Radermacher <blip@mockmoon-cybernetics.ch>
  Copyright (C) 2023 David Phillip Oster <davidphilliposter@gmail.com>

	This program is free software; you can redistribute it and/or modify it
	under the terms of the GNU General Public License version 3 as
	published by the Free Software Foundation
	
	This program is distributed in the hope that it will be useful, but
	WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
	See the GNU General Public License for more details.
	
	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software Foundation,
	Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*/

#include <stdio.h>	/* printf() */
#include <stdlib.h>	/* exit(), malloc() */
#include <stdbool.h>
#include <string.h>	/* strcmp(), memcmp() */
#include <sys/types.h>	/* open() */
#include <sys/stat.h>	/* open() */
#include <fcntl.h>	/* open() */

#ifdef __APPLE__
#include "TargetConditionals.h"
#endif
#if TARGET_OS_MAC // Oster: i.e., is Mac.
#include "gd.h"
#include <locale.h>
#else
#include <gd.h> // oster
#endif
#include "config.h"
#include "gettext.h"	/* gettext(), ngettext() */
#include "ptouch.h"

#define _(s) gettext(s)

#define MAX_LINES 4	/* maybe this should depend on tape size */

gdImage *image_load(const char *file);
void rasterline_setpixel(uint8_t rasterline[16], int pixel);
int get_baselineoffset(char *text, char *font, int fsz);
int find_fontsize(int want_px, char *font, char *text);
int needed_width(char *text, char *font, int fsz);
int print_img(ptouch_dev ptdev, gdImage *im);
int write_png(gdImage *im, const char *file);
gdImage *img_append(gdImage *in_1, gdImage *in_2);
gdImage *img_cutmark(int tape_width);
gdImage *render_text(char *font, char *line[], int lines, int tape_width);
void unsupported_printer(ptouch_dev ptdev);
void usage(char *progname);
int parse_args(int argc, char **argv);

// char *font_file="/usr/share/fonts/TTF/Ubuntu-M.ttf";
// char *font_file="Ubuntu:medium";
char *font_file="Helvetica"; // Oster
char *save_png=NULL;
int verbose=0;
long fontsize=0;  // Oster
bool debug=false;

/* --------------------------------------------------------------------
   -------------------------------------------------------------------- */

void rasterline_setpixel(uint8_t rasterline[16], int pixel)
{
	if (pixel > 128) {
		return;
	}
	rasterline[15-(pixel/8)] |= (uint8_t)(1<<(pixel%8));
	return;
}

int print_img(ptouch_dev ptdev, gdImage *im)
{
	int d,i,k,offset,tape_width;
	uint8_t rasterline[16];

	if (!im) {
		fprintf(stderr, _("nothing to print\n"));
		return -1;
	}
	tape_width=ptouch_getmaxwidth(ptdev);
	/* find out whether color 0 or color 1 is darker */
	d=(gdImageRed(im,1)+gdImageGreen(im,1)+gdImageBlue(im,1) < gdImageRed(im,0)+gdImageGreen(im,0)+gdImageBlue(im,0))?1:0;
	if (gdImageSY(im) > tape_width) {
		fprintf(stderr, _("image is too large (%ipx x %ipx)\n"), gdImageSX(im), gdImageSY(im));
		fprintf(stderr, _("maximum printing width for this tape is %ipx\n"), tape_width);
		return -1;
	}
	offset=64-(gdImageSY(im)/2);	/* always print centered  */
	if ((ptdev->devinfo->flags & FLAG_RASTER_PACKBITS) == FLAG_RASTER_PACKBITS) {
		if (debug) {
			printf("enable PackBits mode\n");
		}
	        ptouch_enable_packbits(ptdev);
	}
	if (ptouch_rasterstart(ptdev) != 0) {
		fprintf(stderr, _("ptouch_rasterstart() failed\n"));
		return -1;
	}
	for (k=0; k<gdImageSX(im); k+=1) {
		memset(rasterline, 0, sizeof(rasterline));
		for (i=0; i<gdImageSY(im); i+=1) {
			if (gdImageGetPixel(im, k, gdImageSY(im)-1-i) == d) {
				rasterline_setpixel(rasterline, offset+i);
			}
		}
		if (ptouch_sendraster(ptdev, rasterline, 16) != 0) {
			fprintf(stderr, _("ptouch_sendraster() failed\n"));
			return -1;
		}
	}
	return 0;
}

/* --------------------------------------------------------------------
	Function	image_load()
	Description	detect the type of a image and try to load it
	Last update	2005-10-16
	Status		Working, should add debug info
   -------------------------------------------------------------------- */

gdImage *image_load(const char *file)
{
	const uint8_t png[8]={0x89,'P','N','G',0x0d,0x0a,0x1a,0x0a};
	char d[10];
	FILE *f;
	gdImage *img=NULL;

	if ((f = fopen(file, "rb")) == NULL) {	/* error cant open file */
		return NULL;
	}
	if (fread(d, sizeof(d), 1, f) != 1) {
		return NULL;
	}
	rewind(f);
	if (memcmp(d, png, 8) == 0) {
		img=gdImageCreateFromPng(f);
	}
	fclose(f);
	return img;
}

int write_png(gdImage *im, const char *file)
{
	FILE *f;

	if ((f = fopen(file, "wb")) == NULL) {
		fprintf(stderr, _("writing image '%s' failed\n"), file);
		return -1;
	}
	gdImagePng(im, f);
	fclose(f);
	return 0;
}

/* --------------------------------------------------------------------
	Find out the difference in pixels between a "normal" char and one
	that goes below the font baseline
   -------------------------------------------------------------------- */
int get_baselineoffset(char *text, char *font, int fsz)
{
	int brect[8];

	if (strpbrk(text, "QgjpqyQµ") == NULL) {	/* if we have none of these */
		return 0;		/* we don't need an baseline offset */
	}				/* else we need to calculate it */
	gdImageStringFT(NULL, &brect[0], -1, font, fsz, 0.0, 0, 0, "o");
	int tmp=brect[1]-brect[5];
	gdImageStringFT(NULL, &brect[0], -1, font, fsz, 0.0, 0, 0, "g");
	return (brect[1]-brect[5])-tmp;
}

/* --------------------------------------------------------------------
	Find out which fontsize we need for a given font to get a
	specified pixel size
	NOTE: This does NOT work for some UTF-8 chars like µ
   -------------------------------------------------------------------- */
int find_fontsize(int want_px, char *font, char *text)
{
	int save=0;
	int brect[8];

	for (int i=4; ; i++) {
		if (gdImageStringFT(NULL, &brect[0], -1, font, i, 0.0, 0, 0, text) != NULL) {
			break;
		}
		if (brect[1]-brect[5] <= want_px) {
			save=i;
		} else {
			break;
		}
	}
	if (save == 0) {
		return -1;
	}
	return save;
}

int needed_width(char *text, char *font, int fsz)
{
	int brect[8];

	if (gdImageStringFT(NULL, &brect[0], -1, font, fsz, 0.0, 0, 0, text) != NULL) {
		return -1;
	}
	return brect[2]-brect[0];
}

gdImage *render_text(char *font, char *line[], int lines, int tape_width)
{
	int brect[8];
	int i, black, x=0, tmp=0, fsz=0;
	char *p;
	gdImage *im=NULL;

	if (debug) {
		printf(_("render_text(): %i lines, font = '%s'\n"), lines, font);
	}
	if (gdFTUseFontConfig(1) != GD_TRUE) {
		fprintf(stderr, _("warning: font config not available\n"));
	}
	if ((int)fontsize > 0 && (int)fontsize == fontsize) {
		fsz= (int)fontsize;
		fprintf(stderr, _("setting font size=%i\n"), fsz);
	} else {
		for (i=0; i<lines; i++) {
			if ((tmp=find_fontsize(tape_width/lines, font, line[i])) < 0) {
				fprintf(stderr, _("could not estimate needed font size\n"));
				return NULL;
			}
			if ((fsz == 0) || (tmp < fsz)) {
				fsz=tmp;
			}
		}
		fprintf(stderr, _("choosing font size=%i\n"), fsz);
	}
	for(i=0; i<lines; i++) {
		tmp=needed_width(line[i], font_file, fsz);
		if (tmp > x) {
			x=tmp;
		}
	}
	im=gdImageCreatePalette(x, tape_width);
	gdImageColorAllocate(im, 255, 255, 255);
	black=gdImageColorAllocate(im, 0, 0, 0);
	/* gdImageStringFT(im,brect,fg,fontlist,size,angle,x,y,string) */
	/* find max needed line height for ALL lines */
	int max_height=0;
	for (i=0; i<lines; i++) {
		if ((p=gdImageStringFT(NULL, &brect[0], -black, font, fsz, 0.0, 0, 0, line[i])) != NULL) {
			fprintf(stderr, _("error in gdImageStringFT: %s\n"), p);
		}
		//int ofs=get_baselineoffset(line[i], font_file, fsz);
		int lineheight=brect[1]-brect[5];
		if (lineheight > max_height) {
			max_height=lineheight;
		}
	}
	if (debug) {
		printf("debug: needed (max) height is %ipx\n", max_height);
	}
	/* now render lines */
	for (i=0; i<lines; i++) {
		int ofs=get_baselineoffset(line[i], font_file, fsz);
		int pos=((i)*(tape_width/(lines)))+(max_height)-ofs-1;
		if (debug) {
			printf("debug: line %i pos=%i ofs=%i\n", i+1, pos, ofs);
		}
		if ((p=gdImageStringFT(im, &brect[0], -black, font, fsz, 0.0, 0, pos, line[i])) != NULL) {
			fprintf(stderr, _("error in gdImageStringFT: %s\n"), p);
		}
	}
	return im;
}

gdImage *img_append(gdImage *in_1, gdImage *in_2)
{
	gdImage *out=NULL;
	int width=0;
	int i_1_x=0;
	int length=0;

	if (in_1 != NULL) {
		width=gdImageSY(in_1);
		length=gdImageSX(in_1);
		i_1_x=gdImageSX(in_1);
	}
	if (in_2 != NULL) {
		length += gdImageSX(in_2);
		/* width should be the same, but let's be sure */
		if (gdImageSY(in_2) > width) {
			width=gdImageSY(in_2);
		}
	}
	if ((width == 0) || (length == 0)) {
		return NULL;
	}
	out=gdImageCreatePalette(length, width);
	if (out == NULL) {
		return NULL;
	}
	gdImageColorAllocate(out, 255, 255, 255);
	gdImageColorAllocate(out, 0, 0, 0);
	if (debug) {
		printf("debug: created new img with size %d * %d\n", length, width);
	}
	if (in_1 != NULL) {
		gdImageCopy(out, in_1, 0, 0, 0, 0, gdImageSX(in_1), gdImageSY(in_1));
		if (debug) {
			printf("debug: copied part 1\n");
		}
	}
	if (in_2 != NULL) {
		gdImageCopy(out, in_2, i_1_x, 0, 0, 0, gdImageSX(in_2), gdImageSY(in_2));
		if (debug) {
			printf("copied part 2\n");
		}
	}
	return out;
}

gdImage *img_cutmark(int tape_width)
{
	gdImage *out=NULL;
	int style_dashed[6];

	out=gdImageCreatePalette(9, tape_width);
	if (out == NULL) {
		return NULL;
	}
	gdImageColorAllocate(out, 255, 255, 255);
	int black=gdImageColorAllocate(out, 0, 0, 0);
	style_dashed[0]=gdTransparent;
	style_dashed[1]=gdTransparent;
	style_dashed[2]=gdTransparent;
	style_dashed[3]=black;
	style_dashed[4]=black;
	style_dashed[5]=black;
	gdImageSetStyle(out, style_dashed, 6);
	gdImageLine(out, 5, 0, 5, tape_width-1, gdStyled);
	return out;
}

gdImage *img_padding(int tape_width, int length)
{
	gdImage *out=NULL;

	if ((length < 1) || (length > 256)) {
		length=1;
	}
	out=gdImageCreatePalette(length, tape_width);
	if (out == NULL) {
		return NULL;
	}
	gdImageColorAllocate(out, 255, 255, 255);
	return out;
}

void usage(char *progname)
{
	fprintf(stderr, "usage: %s [options] <print-command(s)>\n", progname);
	fprintf(stderr, "options:\n");
	fprintf(stderr, "\t--font <file>\t\tuse font <file> or <name>\n");
	fprintf(stderr, "\t--writepng <file>\tinstead of printing, write output to png file\n");
	fprintf(stderr, "\t\t\t\tThis currently works only when using\n\t\t\t\tEXACTLY ONE --text statement\n");
	fprintf(stderr, "print-commands:\n");
	fprintf(stderr, "\t--image <file>\t\tprint the given image which must be a 2 color\n");
	fprintf(stderr, "\t\t\t\t(black/white) png\n");
	fprintf(stderr, "\t--text <text>\t\tPrint 1-4 lines of text.\n");
	fprintf(stderr, "\t\t\t\tIf the text contains spaces, use quotation marks\n\t\t\t\taround it.\n");
	fprintf(stderr, "\t--cutmark\t\tPrint a mark where the tape should be cut\n");
	fprintf(stderr, "\t--fontsize\t\tManually set fontsize\n");
	fprintf(stderr, "\t--pad <n>\t\tAdd n pixels padding (blank tape)\n");
	exit(1);
}

/* here we don't print anything, but just try to catch syntax errors */
int parse_args(int argc, char **argv)
{
	int lines, i;

	for (i=1; i<argc; i++) {
		if (*argv[i] != '-') {
			break;
		}
		if (strcmp(&argv[i][1], "-font") == 0) {
			if (i+1<argc) {
				font_file=argv[++i];
			} else {
				usage(argv[0]);
			}
		} else if (strcmp(&argv[i][1], "-fontsize") == 0) {
			if (i+1<argc) {
				i++;
			} else {
				usage(argv[0]);
			}
		} else if (strcmp(&argv[i][1], "-writepng") == 0) {
			if (i+1<argc) {
				save_png=argv[++i];
			} else {
				usage(argv[0]);
			}
		} else if (strcmp(&argv[i][1], "-cutmark") == 0) {
			continue;	/* not done here */
		} else if (strcmp(&argv[i][1], "-debug") == 0) {
			debug=true;
		} else if (strcmp(&argv[i][1], "-info") == 0) {
			continue;	/* not done here */
		} else if (strcmp(&argv[i][1], "-image") == 0) {
			if (i+1<argc) {
				i++;
			} else {
				usage(argv[0]);
			}
		} else if (strcmp(&argv[i][1], "-pad") == 0) {
			if (i+1<argc) {
				i++;
			} else {
				usage(argv[0]);
			}
		} else if (strcmp(&argv[i][1], "-text") == 0) {
			for (lines=0; (lines < MAX_LINES) && (i < argc); lines++) {
				if ((i+1 >= argc) || (argv[i+1][0] == '-')) {
					break;
				}
				i++;
			}
		} else if (strcmp(&argv[i][1], "-version") == 0) {
			fprintf(stderr, _("ptouch-print by Dominic Radermacher, Mac version %s \n"), VERSION);
			exit(0);
		} else {
			usage(argv[0]);
		}
	}
	return i;
}

int main(int argc, char *argv[])
{
	int i, lines = 0, tape_width;
	char *line[MAX_LINES];
	gdImage *im=NULL;
	gdImage *out=NULL;
	ptouch_dev ptdev=NULL;

	setlocale(LC_ALL, "");
#if TARGET_OS_MAC
#else
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
#endif
	i=parse_args(argc, argv);
	if (i != argc) {
		usage(argv[0]);
	}
	if ((ptouch_open(&ptdev)) < 0) {
		return 5;
	}
	if (ptouch_init(ptdev) != 0) {
		fprintf(stderr, _("ptouch_init() failed\n"));
	}
	if (ptouch_getstatus(ptdev) != 0) {
		fprintf(stderr, _("ptouch_getstatus() failed\n"));
		return 1;
	}
	tape_width=ptouch_getmaxwidth(ptdev);
	for (i=1; i<argc; i++) {
		if (*argv[i] != '-') {
			break;
		}
		if (strcmp(&argv[i][1], "-font") == 0) {
			if (i+1<argc) {
				font_file=argv[++i];
			} else {
				usage(argv[0]);
			}
		} else if (strcmp(&argv[i][1], "-fontsize") == 0) {
			if (i+1<argc) {
				fontsize=strtol(argv[++i], NULL, 10);
			} else {
				usage(argv[0]);
			}
		} else if (strcmp(&argv[i][1], "-writepng") == 0) {
			if (i+1<argc) {
				save_png=argv[++i];
			} else {
				usage(argv[0]);
			}
		} else if (strcmp(&argv[i][1], "-info") == 0) {
			printf(_("maximum printing width for this tape is %ipx\n"), tape_width);
			printf("media type = %02x\n", ptdev->status->media_type);
			printf("media width = %d mm\n", ptdev->status->media_width);
			printf("tape color = %02x\n", ptdev->status->tape_color);
			printf("text color = %02x\n", ptdev->status->text_color);
			printf("error = %04x\n", ptdev->status->error);
			exit(0);
		} else if (strcmp(&argv[i][1], "-image") == 0) {
			im=image_load(argv[++i]);
			out=img_append(out, im);
			gdImageDestroy(im);
			im = NULL;
		} else if (strcmp(&argv[i][1], "-text") == 0) {
			for (lines=0; (lines < MAX_LINES) && (i < argc); lines++) {
				if ((i+1 >= argc) || (argv[i+1][0] == '-')) {
					break;
				}
				i++;
				line[lines]=argv[i];
			}
			if (lines) {
				if ((im=render_text(font_file, line, lines, tape_width)) == NULL) {
					fprintf(stderr, _("could not render text\n"));
					return 1;
				}
				out=img_append(out, im);
				gdImageDestroy(im);
				im = NULL;
			}
		} else if (strcmp(&argv[i][1], "-cutmark") == 0) {
			im=img_cutmark(tape_width);
			out=img_append(out, im);
			gdImageDestroy(im);
			im = NULL;
		} else if (strcmp(&argv[i][1], "-pad") == 0) {
			int length= (int)strtol(argv[++i], NULL, 10);
			im=img_padding(tape_width, length);
			out=img_append(out, im);
			gdImageDestroy(im);
			im = NULL;
		} else if (strcmp(&argv[i][1], "-debug") == 0) {
			debug = true;
		} else {
			usage(argv[0]);
		}
	}
	if (out) {
		if (save_png) {
			write_png(out, save_png);
		} else {
			print_img(ptdev, out);
			if (ptouch_eject(ptdev) != 0) {
				fprintf(stderr, _("ptouch_eject() failed\n"));
				return -1;
			}
		}
		gdImageDestroy(out);
	}
	if (im != NULL) {
		gdImageDestroy(im);
	}
	ptouch_close(ptdev);
	libusb_exit(NULL);
	return 0;
}
